ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/DefaultBoxSelector.scala

```diff
@@ -89,43 +91,45 @@ object DefaultBoxSelector extends BoxSelector {
-                     ): Either[BoxSelectionError, Seq[ErgoBoxAssets]] = {
-    AssetUtils.subtractAssetsMut(foundBoxAssets, targetBoxAssets)
-    val changeBoxesAssets: Seq[mutable.Map[ModifierId, Long]] = foundBoxAssets.grouped(MaxTokens).toSeq
+  /**
+    * This method calculates and creates change box for transaction
+    * or throws specific error if transaction is invalid
+    *
+    * @param foundBalance - Origin balance, balance from which we subtract other amount
+    * @param targetBalance - Subtracted amount
+    * @param foundBoxAssets - Origin asset balances
+    * @param targetBoxAssets - Subtracted asset amounts
+    */
+  def formChangeBox(
+                     foundBalance: Long,
+                     targetBalance: Long,
+                     foundBoxAssets: mutable.Map[ModifierId, Long],
+                     targetBoxAssets: TokensMap
+                   ): Either[BoxSelectionError, Option[ErgoBoxAssets]] = {
     val changeBalance = foundBalance - targetBalance
-    //at least a minimum amount of ERG should be assigned per a created box
-    if (changeBoxesAssets.size * MinBoxValue > changeBalance) {
-      Left(NotEnoughCoinsForChangeBoxesError(
-        s"Not enough ERG $changeBalance to create ${changeBoxesAssets.size} change boxes, \nfor $changeBoxesAssets"
-      ))
-    } else {
-      val changeBoxes = if (changeBoxesAssets.nonEmpty) {
-        val baseChangeBalance = changeBalance / changeBoxesAssets.size
-
-        val changeBoxesNoBalanceAdjusted = changeBoxesAssets.map { a =>
-          ErgoBoxAssetsHolder(baseChangeBalance, a.toMap)
-        }
-
-        val modifiedBoxOpt = changeBoxesNoBalanceAdjusted.headOption.map { firstBox =>
-          ErgoBoxAssetsHolder(
-            changeBalance - baseChangeBalance * (changeBoxesAssets.size - 1),
-            firstBox.tokens
-          )
-        }
-
-        modifiedBoxOpt.toSeq ++ changeBoxesNoBalanceAdjusted.tail
-      } else if (changeBalance > 0) {
-        Seq(ErgoBoxAssetsHolder(changeBalance))
-      } else {
-        Seq.empty
+    foundBoxAssets.foldLeft[Either[BoxSelectionError, TokensMap]](Right(Map.empty)) { case (acc, (id, amount)) =>
+      targetBoxAssets.get(id) match {
+        case Some(value) => acc.map(_.updated(id, amount - value))
+        case None => Left(NoSuchTokensError(
+          s"There are no $id tokens in target"
+        ))
       }
-      Right(changeBoxes)
+      // First, we iterate over found tokens and subtract target
+      // assets from found if token exists, and throw exception otherwise
+    }.flatMap {
+      tMap =>
+        // Check if subtracted ERG amount is greater than balance and throw exception if so
+        if (changeBalance < 0)
+          Left(NotEnoughCoinsForChangeBoxError(s"Not enough ERG $changeBalance to create change box"))
+        // Check if all subtracted token amounts are less than token balances, throw exc if not
+        else if (!tMap.forall { case (_, amount) => amount >= 0 })
```

Comment:
```suggestion
        else if (tMap.exists { case (_, amount) => amount < 0 })
```
should be slightly more efficient